service/src/main/java/com/commercetools/pspadapter/payone/notification/common/CaptureNotificationProcessor.java

```diff
@@ -28,37 +30,64 @@ public CaptureNotificationProcessor(TenantFactory tenantFactory, TenantConfig te
-                .map(transaction -> listBuilder.build())
-                .orElseGet(() -> {
-                    final MonetaryAmount amount = MoneyImpl.of(notification.getPrice(), notification.getCurrency());
+                .map(transaction -> updateChargeTransactionState(transaction, notification, actionsBuilder))
+                .orElseGet(() -> createChargeTransaction(notification, actionsBuilder, sequenceNumber))
+                .build();
+    }
+
+    /**
+     * Add {@link ChangeTransactionState} action to {@code actionsBuilder} if current transaction is still in different
+     * state.
+     */
+    protected ImmutableList.Builder<UpdateAction<Payment>> updateChargeTransactionState(
+            @Nonnull final Transaction transaction,
+            @Nonnull final Notification notification,
+            @Nonnull final ImmutableList.Builder<UpdateAction<Payment>> actionsBuilder) {
+        final TransactionState newTransactionState = notification.getTransactionStatus().getCtTransactionState();
+        if (transaction.getState() != newTransactionState) {
+            actionsBuilder.add(ChangeTransactionState.of(newTransactionState, transaction.getId()));
+        }
+
+        return actionsBuilder;
+    }
+
+    /**
+     * For payments except {@code BANK_TRANSFER-ADVANCE} ({@link ClearingType#PAYONE_VOR}) add {@link AddTransaction}
+     * update action with {@link TransactionType#CHARGE} and state equal to
+     * {@link io.sphere.sdk.payments.TransactionState Notification#getTransactionStatus()#getCtTransactionState()}
+     */
+    protected ImmutableList.Builder<UpdateAction<Payment>> createChargeTransaction(
+            @Nonnull final Notification notification,
+            @Nonnull final ImmutableList.Builder<UpdateAction<Payment>> actionsBuilder,
+            @Nonnull final String sequenceNumber) {
+        final MonetaryAmount amount = MoneyImpl.of(notification.getPrice(), notification.getCurrency());
 
-                    //For all payment methods that use PayOne preauthorization and capture but only one CTP-charge
-                    //we must not create a second charge transaction
-                    //at the moment that is only "BANK_TRANSFER_ADVANCE"
-                    if (ClearingType.PAYONE_VOR != ClearingType.getClearingTypeByCode(notification.getClearingtype())) {
-                    listBuilder.add(AddTransaction.of(TransactionDraftBuilder.of(TransactionType.CHARGE, amount)
-                            .timestamp(toZonedDateTime(notification))
-                            .state(TransactionState.PENDING)
-                            .interactionId(sequenceNumber)
-                            .build()));
-                    }
+        //For all payment methods that use PayOne preauthorization and capture but only one CTP-charge
+        //we must not create a second charge transaction
+        //at the moment that is only "BANK_TRANSFER_ADVANCE"
```

Comment:
Is this comment still valid?

Reply:
I can't relate it to the code change in the if condition below..

Reply:
the comment is still valid, it is related to this condition:
```java 
ClearingType.PAYONE_VOR != ClearingType.getClearingTypeByCode(notification.getClearingtype())
```

But this might be re-factored in the future cos we also not sure this what we need.

Reply:
@andrii-kovalenko-ct `For all payment methods that use PayOne preauthorization and capture but only one CTP-charge` I still don't see this comment related to the condition..

Reply:
i'm trying to write is as a pseudocode:
```java
if (this is a capture notification) { // because this is CaptureNotificationProcessor class
  if (paymentMethod is BANK_TRANSFER_ADVANCE) {
    do nothing (don't add a transaction, because it will be second CHARGE transaction in this payment method)
  } else { // all other payment methods
    add CHARGE transaction
  }
}

Reply:
is `PAYONE_VOR` (in the actual code) the same as `BANK_TRANSFER_ADVANCE` (in the pseudocode)?

Also where is the `if (this is a capture notification) {` in the actual code?

Reply:
> is PAYONE_VOR (in the actual code) the same as BANK_TRANSFER_ADVANCE (in the pseudocode)?

yes, it's enum value: https://github.com/commercetools/commercetools-payone-integration/blob/d59b48945c6b9777fe3f43ecc9cc491fd17cd52a/service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/ClearingType.java#L19

> Also where is the if (this is a capture notification) in the actual code?

like written in the the comment to that line: _because this is **Capture**NotificationProcessor class_, and this means all the logic implemented in this class is executed exclusively in context of `capture` notification processing. I put that `if` into the pseudocode snippet only to explain "_and capture_" words from the comment

Reply:
ok thanks for the explanation.